src/cdk-experimental/menu/context-menu.ts

```diff
@@ -0,0 +1,318 @@
+      event.stopPropagation();
+
+      this._contextMenuTracker.update(this);
+      this.openMenu({x: event.clientX, y: event.clientY});
+    }
+  }
+
+  /** Whether the attached menu is open. */
+  isMenuOpen() {
+    return !!this._overlayRef?.hasAttached();
+  }
+
+  /**
+   * Get the configuration object used to create the overlay.
+   * @param coordinates the location to place the opened menu
+   */
+  private _getOverlayConfig(coordinates: ContextMenuCoordinates) {
+    return new OverlayConfig({
+      positionStrategy: this._getOverlayPositionStrategy(coordinates),
+      scrollStrategy: this._overlay.scrollStrategies.block(),
+      direction: this._directionality,
+    });
+  }
+
+  /**
+   * Build the position strategy for the overlay which specifies where to place the menu.
+   * @param coordinates the location to place the opened menu
+   */
+  private _getOverlayPositionStrategy(
+    coordinates: ContextMenuCoordinates
+  ): FlexibleConnectedPositionStrategy {
+    return this._overlay
+      .position()
+      .flexibleConnectedTo(coordinates)
+      .withDefaultOffsetX(this._options.menuOffsetX)
+      .withDefaultOffsetY(this._options.menuOffsetY)
+      .withPositions(this._getOverlayPositions());
+  }
+
+  /**
+   * Determine and return where to position the opened menu relative to the mouse location.
+   */
+  private _getOverlayPositions(): ConnectedPosition[] {
+    return [
+      {originX: 'end', originY: 'top', overlayX: 'start', overlayY: 'top'},
+      {originX: 'start', originY: 'top', overlayX: 'end', overlayY: 'top'},
+      {originX: 'end', originY: 'bottom', overlayX: 'start', overlayY: 'bottom'},
+      {originX: 'start', originY: 'bottom', overlayX: 'end', overlayY: 'bottom'},
+    ];
+  }
+
+  /**
+   * Get the portal to be attached to the overlay which contains the menu. Allows for the menu
+   * content to change dynamically and be reflected in the application.
+   */
+  private _getPortal() {
+    const hasMenuContentChanged = this.menuPanel._templateRef !== this._panelContent?.templateRef;
+    if (this.menuPanel && (!this._panelContent || hasMenuContentChanged)) {
+      this._panelContent = new TemplatePortal(this.menuPanel._templateRef, this._viewContainerRef);
+    }
+
+    return this._panelContent;
+  }
+
+  /**
+   * Subscribe to the document click and context menu events and close out the menu when emitted.
+   */
+  private _setCloseListener() {
+    merge(fromEvent(this._document, 'click'), fromEvent(this._document, 'contextmenu'))
+      .pipe(take(1), takeUntil(this._stopDocumentListener))
+      .subscribe(event => {
+        // stop the default context menu from appearing if user right-clicked somewhere outside of
+        // any context menu directive or if a user right-clicked inside of the opened menu and just
+        // close it.
+        if (event.type === 'contextmenu') {
+          event.preventDefault();
+          this.closeMenu();
```

Comment:
I'm not sure this bit is necessary- I tested a few other Google apps (gmail, docs, calendar) and they don't seem to prevent native context menus from opening when you click somewhere without an open one (while you have an open custom menu)

Reply:
Do we want to prevent the default context menu from opening when right clicked inside a custom context menu?

Reply:
Hmm, interesting- gmail completely ignores right-clicks inside of a custom context menu while Google Docs and Calendar both just treat them as normal clicks. I would probably do what gmail does here since it seems the most predictable.

Reply:
Agreed. 

So I've implemented it as:

- if regular click occurs outside of the open context menu, close the open menu
- if right click occurs inside the context menu, do nothing
- if right click occurs outside of the context menu (and outside the context) close the menu and either open the new context menu or the regular context menu